tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/mhlo_mark_shape_calc.cc

```diff
@@ -0,0 +1,313 @@
+      // placed on cpu.
+      SmallVector<Attribute, 4> attrs(tp.size(), builder.getBoolAttr(true));
+      op->setAttr(kDiscShapeCalcAttr, ArrayAttr::get(tp.getContext(), attrs));
+    } else {
+      op->setAttr(kDiscShapeCalcAttr, builder.getBoolAttr(true));
+    }
+  }
+}
+
+// Regard any mhlo Ops that calculates i32 as shape Ops. This is an rule based
+// optimization that mimicking the behavior of tensorflow
+void MarkShapeCalc::MarkRegardAsShapeCalcOps() {
+  ModuleOp module = getOperation();
+  Builder builder(&getContext());
+
+  module.walk([&](Operation* op) {
+    if (!isMhloDialect(op)) return;
+
+    if (isa<mhlo::TupleOp, mhlo::GetTupleElementOp, mhlo::WhileOp, mhlo::IfOp,
+            mhlo::ReturnOp>(op)) {
+      return;
+    }
+    // Skip the Op that is already marked shape Op
+    auto attr = op->getAttrOfType<BoolAttr>(kDiscShapeCalcAttr);
+    if ((attr != nullptr) && (attr.getValue() == true)) return;
+
+    if (isa<mhlo::GetDimensionSizeOp, mhlo::PrintOp>(op)) {
+      op->setAttr(kDiscShapeCalcAttr, builder.getBoolAttr(true));
+      return;
+    }
+
+    // Ops that only cares about the output element type
+    if (isa<mhlo::ConstOp, mhlo::SelectOp, mhlo::IotaOp, mhlo::DynamicIotaOp>(
+            op)) {
+      auto result_ty = op->getResult(0).getType().dyn_cast<RankedTensorType>();
+      assert(result_ty && "unexpected non ranked type for ConstOp");
+      auto elem_type = result_ty.getElementType();
+      if (elem_type.isInteger(32)) {
+        op->setAttr(kDiscShapeCalcAttr, builder.getBoolAttr(true));
+      }
+      return;
+    }
+
+    StringRef op_name = op->getName().getStringRef();
+    bool is_shape_calc_op = false;
+    // Follow the rule of kPlaceRuleMap exist, or else follow
+    // kShapeCalcOperandMap
+    auto it = kPlaceRuleMap.find(op_name);
+    if (it != kPlaceRuleMap.end()) {
+      auto dominant_idx = it->second;
+      auto operand_ty =
+          op->getOperand(dominant_idx).getType().dyn_cast<RankedTensorType>();
+      assert(operand_ty && "unexpected non unranked type of operand");
+      if (operand_ty.getElementType().isInteger(32)) {
+        is_shape_calc_op = true;
+      }
+    } else {
+      DenseSet<int> shape_operand_indices;
+      auto iter = kShapeCalcOperandMap.find(op_name);
+      if (iter != kShapeCalcOperandMap.end()) {
+        shape_operand_indices = iter->second;
+      }
+      for (int idx = 0; idx < op->getNumOperands(); ++idx) {
+        // if it is not "shape operand", then it must be "data operand"
+        if (shape_operand_indices.find(idx) == shape_operand_indices.end()) {
+          auto operand_ty =
+              op->getOperand(idx).getType().dyn_cast<RankedTensorType>();
+          if (!operand_ty) continue;
+          auto elem_type = operand_ty.getElementType();
+          if (elem_type.isInteger(32)) {
+            is_shape_calc_op = true;
+            break;
+          }
+        }
+      }
```

Comment:
Seems like this could be more efficient:
```
       auto iter = kShapeCalcOperandMap.find(op_name);
       if (iter != kShapeCalcOperandMap.end()) {
         auto &shape_operand_indices = iter->second;
         for (int idx : shape_operand_indices) {
           auto operand_ty =
               op->getOperand(idx).getType().dyn_cast<RankedTensorType>();
           if (!operand_ty) continue;
           auto elem_type = operand_ty.getElementType();
           if (elem_type.isInteger(32)) {
             is_shape_calc_op = true;
             break;
           }
         }
       }
```

Reply:
Done